Vertical margin proposal

joebourne opened the pull request.

cutandpastey commented on proposals/vertical-rhythm.md:
```diff
@@ -0,0 +1,97 @@
+function rhythm(size, direction = 'bottom') {
+  try {
+    if (!['top', 'bottom'].includes(direction)) {
+      throw `direction must be either 'top' or 'bottom'`;
```
Probably want to throw an actual error here, as it has a stack.

Joebourne replied:
I've redone the error handling as a whole. Please can you check I've got the right idea?

cutandpastey commented on proposals/vertical-rhythm.md:
````diff
@@ -0,0 +1,97 @@
+
+```js
+type Props = {|
+  variant: 'section' | 'semiSection',
````
Camel case or kebab case?

Joebourne replied:
👍 will go with kebab

cutandpastey commented on proposals/vertical-rhythm.md:
````diff
@@ -0,0 +1,97 @@
+}
+```
+
+As for the `<Section />` component, it will take a boolean prop `isFirst` that will determine whether or not the margin should be added to top and bottom.
````
I thought we agreed you could assign both ways?:

```js
<Section spacingTop="section" spacingBottom="semi-section" />
```

I dont like the naming of `spacingTop` % `spacingBottom` so am open to suggestions?

Joebourne replied:
went with `marginTop` and `marginBottom` as that's explicit.

richiemccoll commented on proposals/vertical-rhythm.md:
````diff
@@ -0,0 +1,97 @@
+The `spacing` module will be renamed and will allow an optional second argument to specify the direction of the margin.
+
+```js
+function rhythm(size, direction = 'bottom') {
````
I think this name is worse than spacing tbh. One of the problems states is that spacing is a confusing name because it could imply space in any direction, you've added that functionality here so surely the name makes sense now?

rlengvenis replied:
or it can by renamed to [vSpacing | vspacing] - to make it more specific

cutandpastey replied:
Its supposed to map to vertical rhythm rather than spacing. Maybe we should just name it that?

Joebourne replied:
gonna go with `verticalMargin` so there can be no confusion

dbbk commented:
# Rhythm
I agree that the spacing util is very restrictive, and the name is confusing. There are certainly places in the UI on a micro level where it makes sense to be setting the margin on the top rather than the bottom, and this has been achieved in some places by duplicating the mixin's logic, which is not ideal.

Enabling `margin-top` as an option is definitely a minimum requirement, and I think this proposal is a good way to achieve that. Although I'm not sure that `5px` makes sense as a default.

This proposal does have a shortcoming though in that it doesn't tackle setting padding. I personally think the CSS variables approach would still be the best way to resolve this fully. That way, on a component's styles you could set;

`margin-bottom: var(--spacing-sm);`
`padding-top: var(--spacing-md);`

etc, without the need for a mixin, and the value of the variable would be dynamically updated in a root media query, so the individual components don't need to know about media queries themselves. Only IE 11 does not support this dynamic variable updating natively, but since it's a desktop browser we could easily transpile it to always use the large versions.

# Section
I'm not sure I understand the correlation between `isFirst` and setting spacing on the top. Also, my initial thought when seeing the `isFirst` prop was, "why can't that be a pseudo selector"?

Joebourne commented:
Thanks @dbbk 

As discussed in the components working group on Monday, we're going to improve things incrementally and stage 1 is to sort out the issue with vertical margin.

Next step is to look more closely at layout components or utilities that concern all types of spacing: margin and padding, both vertical and horizontal.

As for CSS variables, this is again something worth looking into but out of the scope of this initial phase.

With regard to your other feedback, I've removed the `isFirst` logic so that should be clearer now, and I agree that the default value from the original spacing utility makes no sense so I've removed that.

dbbk commented:
Ah fair enough, I wasn’t aware there was a component group meeting. Whoever
runs it please invite me in future.

On Thu, 23 Aug 2018 at 9:40 am, Joe Bourne <notifications@github.com> wrote:

> Thanks @dbbk <https://github.com/dbbk>
>
> As discussed in the components working group on Monday, we're going to
> improve things incrementally and stage 1 is to sort out the issue with
> vertical margin.
>
> Next step is to look more closely at layout components or utilities that
> concern all types of spacing: margin and padding, both vertical and
> horizontal.
>
> As for CSS variables, this is again something worth looking into but out
> of the scope of this initial phase.
>
> With regard to your other feedback, I've removed the isFirst logic so
> that should be clearer now, and I agree that the default value from the
> original spacing utility makes no sense so I've removed that.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/festicket/front-end-meta/pull/16#issuecomment-415338779>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AArcFBVlloGeqBgFjOm6wRXzDWZDKkWTks5uTmqFgaJpZM4WHIN7>
> .
>
